abci/types/application.go

```diff
@@ -5,34 +5,32 @@ import "context"
 //go:generate ../../scripts/mockery_generate.sh Application
 // Application is an interface that enables any finite, deterministic state machine
 // to be driven by a blockchain-based replication engine via the ABCI.
-// All methods take a RequestXxx argument and return a ResponseXxx argument,
-// except CheckTx/DeliverTx, which take `tx []byte`, and `Commit`, which takes nothing.
 type Application interface {
 	// Info/Query Connection
-	Info(context.Context, RequestInfo) ResponseInfo    // Return application info
-	Query(context.Context, RequestQuery) ResponseQuery // Query for state
+	Info(context.Context, RequestInfo) (*ResponseInfo, error)    // Return application info
```

Comment:
What's the benefit of having the responses be references to structs, but the requests not?

Reply:
One (I?) want return values to be nil, so in the error case you can just return nil. For the argument struct--in most cases--you can construct the inline to the function argument, and it's not a big deal. I'm not worried about the copy (struct values) or the heap allocation (pointers), particularly, just the semantics. 

One of the big changes, (and I think useful parallels) at this point is that the Client interface and the application interface are now the same. (basically).

Reply:
Adding on to @tychoish's answer: These types are mostly protobuf generated types, and the generated methods are all on the pointer anyway. Although there's a lot of existing use where we pass these around by value, that's generally _not_ recommended for protobuf types. Apart from the costs of shipping around potentially very large values (proto structs are big and tend to grow over time), passing by value forces us to do some shenanigans when we want to handle them (e.g., adding/removing pointers).

Ideally I'd like to fix _all_ the places we pass protobuf types by value, but these are a good start (and since it's new API surface, it won't cause any major issues).

Reply:
I suppose my question is, why not just pass the requests by reference then too?

Reply:
> I suppose my question is, why not just pass the requests by reference then too?

I'm in favour of it. 🙂

Reply:
> I'm in favour of it. :slightly_smiling_face:

Same.

Reply:
we can do that in another PR, but I think we should change the client API in that case.